pstoreds/protobook.go

```diff
@@ -0,0 +1,122 @@
+package pstoreds
+
+import (
+	"fmt"
+	"sync"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+)
+
+type dsProtoBook struct {
+	lks  [256]sync.RWMutex
+	meta pstore.PeerMetadata
+}
+
+var _ pstore.ProtoBook = (*dsProtoBook)(nil)
+
+func NewProtoBook(meta pstore.PeerMetadata) pstore.ProtoBook {
+	return &dsProtoBook{meta: meta}
+}
+
+func (pb *dsProtoBook) Lock(p peer.ID) {
+	b := []byte(p)
+	pb.lks[b[len(b)-1]].Lock()
+}
+
+func (pb *dsProtoBook) Unlock(p peer.ID) {
+	b := []byte(p)
+	pb.lks[b[len(b)-1]].Unlock()
+}
+
+func (pb *dsProtoBook) RLock(p peer.ID) {
+	b := []byte(p)
+	pb.lks[b[len(b)-1]].RLock()
+}
+
+func (pb *dsProtoBook) RUnlock(p peer.ID) {
+	b := []byte(p)
+	pb.lks[b[len(b)-1]].RUnlock()
+}
+
+func (pb *dsProtoBook) SetProtocols(p peer.ID, protos ...string) error {
+	pb.Lock(p)
+	defer pb.Unlock(p)
```

Comment:
We could also just require a transaction datastore and avoid locking entirely. cc @raulk?